relm4-components/src/simple_adw_combo_row.rs

```diff
@@ -0,0 +1,121 @@
+/// The only requirement is that the inner type implements [`ToString`] and [`Debug`].
+///
+/// To get notified when the selection changed, you can use
+/// [`Connector::forward()`](relm4::component::Connector::forward())
+/// after launching the component.
+pub struct SimpleComboRow<E: ToString> {
+    /// The variants that can be selected.
+    pub variants: Vec<E>,
+    /// The index of the active element or [`None`] is nothing is selected.
+    pub active_index: Option<usize>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+/// The message type of [`SimpleComboRow`].
+pub enum SimpleComboRowMsg<E: ToString> {
+    /// Overwrite the current values.
+    UpdateData(SimpleComboRow<E>),
+    /// Set the index of the active element.
+    SetActiveIdx(usize),
+    #[doc(hidden)]
+    UpdateIndex(usize),
+}
+
+impl<E> Component for SimpleComboRow<E>
+where
+    E: ToString + 'static + Debug,
+{
+    type CommandOutput = ();
+    type Input = SimpleComboRowMsg<E>;
+    type Output = usize;
+    type Init = Self;
+    type Root = adw::ComboRow;
+    type Widgets = adw::ComboRow;
+
+    fn init_root() -> Self::Root {
+        adw::ComboRow::default()
+    }
+
+    fn init(
+        model: Self::Init,
+        root: &Self::Root,
+        sender: ComponentSender<Self>,
+    ) -> ComponentParts<Self> {
+        let widgets = root.clone();
+
+        model.render(&widgets);
+
+        widgets.connect_selected_notify(move |combo_box| {
+            sender.input(Self::Input::UpdateIndex(combo_box.selected() as _));
+        });
+
+        ComponentParts { model, widgets }
+    }
+
+    fn update_with_view(
+        &mut self,
+        widgets: &mut Self::Widgets,
+        input: Self::Input,
+        sender: ComponentSender<Self>,
+        _root: &Self::Root,
+    ) {
+        match input {
+            SimpleComboRowMsg::UpdateIndex(idx) => {
+                // Ignore send errors because the component might
+                // be detached.
+                sender.output(idx).ok();
+                self.active_index = Some(idx);
+            }
+            SimpleComboRowMsg::SetActiveIdx(idx) => {
+                if idx < self.variants.len() {
+                    self.active_index = Some(idx);
+                    widgets.set_selected(idx as u32);
+                }
+            }
+            SimpleComboRowMsg::UpdateData(data) => {
+                *self = data;
+                self.render(widgets);
+            }
+        }
+    }
+}
+
+impl<E> SimpleComboRow<E>
+where
+    E: ToString,
+{
+    fn render(&self, combo_box: &adw::ComboRow) {
+        use std::ops::Deref;
+        if combo_box.model().is_some() {
+            unreachable!()
+        }
+
+        let variants: Vec<String> = self.variants.iter().map(ToString::to_string).collect();
+        let variants: Vec<&str> = variants.iter().map(String::deref).collect();
+        combo_box.set_model(Some(&StringList::new(&variants)));
```

Comment:
You should be able to directly collect into the `StringList` in just one line: https://gtk-rs.org/gtk4-rs/git/docs/gtk4/struct.StringList.html#impl-FromIterator%3CString%3E-for-StringList

Reply:
Oh, that's much simpler and performant thanks.

It's crazy how far the GTK actually go!